

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genepi.GenEpi &mdash; GenEpi  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> GenEpi
          

          
          </a>

          
            
            
              <div class="version">
                2.0.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../format.html">I/O File Fomats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../example.html">More Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../workflow.html">How it Work</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API Documentations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Release History</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">GenEpi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>genepi.GenEpi</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genepi.GenEpi</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Created on Apr 2019</span>

<span class="sd">@author: Chester (Yu-Chuan Chang)</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="c1"># import libraries</span>
<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span> <span class="k">as</span> <span class="nn">mp</span>
<span class="kn">import</span> <span class="nn">genepi</span>

<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="c1"># define functions </span>
<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<div class="viewcode-block" id="ArgumentsParser"><a class="viewcode-back" href="../../api.html#genepi.GenEpi.ArgumentsParser">[docs]</a><span class="k">def</span> <span class="nf">ArgumentsParser</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    To obtain and parse the arguments from user.</span>

<span class="sd">    Args:</span>
<span class="sd">        None</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.ArgumentParser</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">### define arguments</span>
    <span class="n">str_description</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="s1">&#39;GenEpi is a package to uncover epistasis associated with phenotypes by a machine learning approach, &#39;</span>
    <span class="s1">&#39;developed by Yu-Chuan Chang at c4Lab of National Taiwan University. &#39;</span>
    <span class="s1">&#39;Github: https://github.com/Chester75321/GenEpi &#39;</span>
    <span class="s1">&#39;Reference: GenEpi: Gene-based Epistasis Discovery Using Machine Learning &#39;</span>
    <span class="s1">&#39;(https://www.biorxiv.org/content/10.1101/421719v5) &#39;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">prog</span><span class="o">=</span><span class="s1">&#39;GenEpi&#39;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">str_description</span><span class="p">)</span>
    
    <span class="c1">### define arguments for I/O</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-g&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;filename of the input .gen file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-p&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;filename of the input phenotype&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-s&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;self-defined genome regions&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output file path&quot;</span><span class="p">)</span>
    
    <span class="c1">### define arguments for modeling</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-m&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">],</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;choose model type: c for classification; r for regression&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-k&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;k of k-fold cross validation&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">(),</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;number of threads&quot;</span><span class="p">)</span>
    
    <span class="c1">### define arguments for step1_downloadUCSCDB</span>
    <span class="n">parser_group_1</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;update UCSC database&quot;</span><span class="p">)</span>
    <span class="n">parser_group_1</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--updatedb&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;enable this function&quot;</span><span class="p">)</span>
    <span class="n">parser_group_1</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-b&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;hg19&quot;</span><span class="p">,</span> <span class="s2">&quot;hg38&quot;</span><span class="p">],</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;human genome build&quot;</span><span class="p">)</span>
    
    <span class="c1">### define arguments for step2_estimateLD</span>
    <span class="n">parser_group_2</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s2">&quot;compress data by LD block&quot;</span><span class="p">)</span>
    <span class="n">parser_group_2</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--compressld&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;enable this function&quot;</span><span class="p">)</span>
    <span class="n">parser_group_2</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-d&quot;</span><span class="p">,</span> <span class="n">required</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;threshold for compression: D prime&quot;</span><span class="p">)</span>
    <span class="n">parser_group_2</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-r&quot;</span><span class="p">,</span> <span class="n">required</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;threshold for compression: R square&quot;</span><span class="p">)</span>
 
    <span class="k">return</span> <span class="n">parser</span></div>

<div class="viewcode-block" id="InputChecking"><a class="viewcode-back" href="../../api.html#genepi.GenEpi.InputChecking">[docs]</a><span class="k">def</span> <span class="nf">InputChecking</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_inputFileName_phenotype</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    To check the numbers of sample are consistent in genotype and phenotype data.</span>

<span class="sd">    Args:</span>
<span class="sd">        str_inputFileName_genotype (str): File name of input genotype data</span>
<span class="sd">        str_inputFileName_phenotype (str): File name of input phenotype data</span>

<span class="sd">    Returns:</span>
<span class="sd">        (tuple): tuple containing:</span>

<span class="sd">            - int_num_genotype (int): The sample number of genotype data</span>
<span class="sd">            - int_num_phenotype (int): The sample number of phenotype data </span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">### check file name exist</span>
    <span class="k">if</span> <span class="n">str_inputFileName_genotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s2">&quot;There is no input genotype file.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">str_inputFileName_phenotype</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s2">&quot;There is no input phenotype file.&quot;</span><span class="p">)</span>
    
    <span class="c1">### count lines of input files</span>
    <span class="n">int_num_genotype</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">))</span>
    <span class="n">int_num_phenotype</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">str_inputFileName_phenotype</span><span class="p">))</span>
    
    <span class="c1">### count sample num. in genotype file</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_inputFile</span><span class="p">:</span>
        <span class="n">list_line</span> <span class="o">=</span> <span class="n">file_inputFile</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
        <span class="n">int_num_genotype_sample</span> <span class="o">=</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">list_line</span><span class="p">)</span> <span class="o">-</span> <span class="mi">5</span><span class="p">)</span> <span class="o">/</span> <span class="mi">3</span>
        <span class="k">if</span> <span class="n">int_num_genotype_sample</span> <span class="o">!=</span> <span class="n">int_num_phenotype</span><span class="p">:</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="s2">&quot;The number of samples in genotype file does not match the number of samples in phenotype file.&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">int_num_genotype</span><span class="p">,</span> <span class="n">int_num_phenotype</span></div>

<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="c1"># main function</span>
<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../api.html#genepi.GenEpi.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    Main function for obtaining user arguments, controling workflow and recording log file.</span>

<span class="sd">    Args:</span>
<span class="sd">        None</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">### obtain arguments from argument parser</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">ArgumentsParser</span><span class="p">()</span><span class="o">.</span><span class="n">parse_args</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    
    <span class="c1">### get arguments for I/O</span>
    <span class="n">str_inputFileName_genotype</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">g</span>
    <span class="n">str_inputFileName_phenotype</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">p</span>
    <span class="n">str_inputFileName_regions</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">s</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">str_inputFileName_regions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">s</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">str_inputFileName_regions</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span>
    <span class="n">str_outputFilePath</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">o</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">str_outputFilePath</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">o</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">str_outputFilePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">)</span>
    <span class="n">int_thread</span> <span class="o">=</span> <span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">t</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">t</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">():</span>
            <span class="n">int_thread</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">t</span><span class="p">)</span>
        
    <span class="k">if</span> <span class="n">str_inputFileName_genotype</span> <span class="o">==</span> <span class="s2">&quot;example&quot;</span> <span class="ow">and</span> <span class="n">str_inputFileName_phenotype</span> <span class="o">==</span> <span class="s2">&quot;example&quot;</span><span class="p">:</span>
        <span class="n">str_command</span> <span class="o">=</span> <span class="s2">&quot;cp &quot;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">genepi</span><span class="o">.</span><span class="vm">__file__</span><span class="p">),</span> <span class="s2">&quot;example&quot;</span><span class="p">,</span> <span class="s2">&quot;sample.csv&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">str_outputFilePath</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">str_command</span><span class="p">)</span>
        <span class="n">str_command</span> <span class="o">=</span> <span class="s2">&quot;cp &quot;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">genepi</span><span class="o">.</span><span class="vm">__file__</span><span class="p">),</span> <span class="s2">&quot;example&quot;</span><span class="p">,</span> <span class="s2">&quot;sample.gen&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">str_outputFilePath</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">str_command</span><span class="p">)</span>
        <span class="n">str_command</span> <span class="o">=</span> <span class="s2">&quot;GenEpi -g &quot;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;sample.gen&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="n">str_command</span> <span class="o">+=</span> <span class="s2">&quot;-p &quot;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;sample.csv&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="n">str_command</span> <span class="o">+=</span> <span class="s2">&quot;-o &quot;</span> <span class="o">+</span> <span class="n">str_outputFilePath</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="n">str_command</span> <span class="o">+=</span> <span class="s2">&quot;--updatedb --compressld&quot;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">str_command</span><span class="p">)</span>
        <span class="k">return</span>
    
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;GenEpi_Log_&quot;</span> <span class="o">+</span> <span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">-%H%M&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">localtime</span><span class="p">())</span> <span class="o">+</span> <span class="s2">&quot;.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_outputFile</span><span class="p">:</span>
        <span class="c1">### create log</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;start analysis at: &quot;</span> <span class="o">+</span> <span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">-%H:%M:%S&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">localtime</span><span class="p">())</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    
        <span class="c1">### log arguments</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Arguments in effect:&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-g (input genotype filename): &quot;</span> <span class="o">+</span> <span class="n">str_inputFileName_genotype</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-p (input phenotype filename): &quot;</span> <span class="o">+</span> <span class="n">str_inputFileName_phenotype</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-s (self-defined genome regions): &quot;</span> <span class="o">+</span> <span class="n">str_inputFileName_regions</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-o (output filepath): &quot;</span> <span class="o">+</span> <span class="n">str_outputFilePath</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-m (model type): &quot;</span> <span class="o">+</span> <span class="s2">&quot;Classification&quot;</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">m</span><span class="o">==</span><span class="s2">&quot;c&quot;</span> <span class="k">else</span> <span class="s2">&quot;Regression&quot;</span>  <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-k (k-fold cross validation): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-t (number of threads): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_thread</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;--updatedb (enable function of update UCSC database): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">updatedb</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-b (human genome build): &quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">b</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;--compressld (enable function of LD data compression): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">compressld</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-d (D prime threshold): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">d</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;-r (R square threshold): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">r</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="c1">### check input format</span>
        <span class="n">int_num_genotype</span><span class="p">,</span> <span class="n">int_num_phenotype</span> <span class="o">=</span> <span class="n">InputChecking</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_inputFileName_phenotype</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Number of variants: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_num_genotype</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Number of samples: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_num_phenotype</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of variants: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_num_genotype</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of samples: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_num_phenotype</span><span class="p">))</span>
        
        <span class="c1">### step1_downloadUCSCDB</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">updatedb</span><span class="p">:</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">DownloadUCSCDB</span><span class="p">(</span><span class="n">str_hgbuild</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">b</span><span class="p">)</span>
    
        <span class="c1">### step2_estimateLD</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compressld</span><span class="p">:</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">EstimateLDBlock</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="o">=</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">float_threshold_DPrime</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">d</span><span class="p">),</span> <span class="n">float_threshold_RSquare</span><span class="o">=</span><span class="nb">float</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">r</span><span class="p">))</span>
            <span class="n">str_inputFileName_genotype</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.gen&quot;</span><span class="p">,</span> <span class="s2">&quot;_LDReduced.gen&quot;</span><span class="p">)))</span>
        
        <span class="c1">### step3_splitByGene</span>
        <span class="k">if</span> <span class="n">str_inputFileName_regions</span> <span class="o">==</span> <span class="s2">&quot;None&quot;</span><span class="p">:</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">SplitByGene</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;snpSubsets&quot;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">SplitByGene</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_inputFileName_UCSCDB</span><span class="o">=</span><span class="n">str_inputFileName_regions</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;snpSubsets&quot;</span><span class="p">))</span>
        
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">m</span><span class="o">==</span><span class="s2">&quot;c&quot;</span><span class="p">:</span>
            <span class="c1">### step4_singleGeneEpistasis_Logistic (for case/control trial)</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">BatchSingleGeneEpistasisLogistic</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;snpSubsets&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="c1">### step5_crossGeneEpistasis_Logistic (for case/control trial)</span>
            <span class="n">float_score_train</span><span class="p">,</span> <span class="n">float_score_test</span> <span class="o">=</span> <span class="n">genepi</span><span class="o">.</span><span class="n">CrossGeneEpistasisLogistic</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;singleGeneResult&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Overall genetic feature performance (F1 score)&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Training: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_train</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Testing (&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-fold CV): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_test</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1">### step6_ensembleWithCovariates (for case/control trial)</span>
            <span class="n">float_score_train</span><span class="p">,</span> <span class="n">float_score_test</span> <span class="o">=</span> <span class="n">genepi</span><span class="o">.</span><span class="n">EnsembleWithCovariatesClassifier</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;crossGeneResult&quot;</span><span class="p">,</span> <span class="s2">&quot;Feature.csv&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Ensemble with co-variate performance (F1 score)&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Training: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_train</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Testing (&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-fold CV): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_test</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">### step4_singleGeneEpistasis_Lasso (for quantitative trial)</span>
            <span class="n">genepi</span><span class="o">.</span><span class="n">BatchSingleGeneEpistasisLasso</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;snpSubsets&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="c1">### step5_crossGeneEpistasis_Lasso (for quantitative trial)</span>
            <span class="n">float_score_train</span><span class="p">,</span> <span class="n">float_score_test</span> <span class="o">=</span> <span class="n">genepi</span><span class="o">.</span><span class="n">CrossGeneEpistasisLasso</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;singleGeneResult&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Overall genetic feature performance (Average of the Pearson and Spearman correlation)&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Training: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_train</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Testing (&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-fold CV): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_test</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="c1">### step6_ensembleWithCovariates (for quantitative trial)</span>
            <span class="n">float_score_train</span><span class="p">,</span> <span class="n">float_score_test</span> <span class="o">=</span> <span class="n">genepi</span><span class="o">.</span><span class="n">EnsembleWithCovariatesRegressor</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="s2">&quot;crossGeneResult&quot;</span><span class="p">,</span> <span class="s2">&quot;Feature.csv&quot;</span><span class="p">),</span> <span class="n">str_inputFileName_phenotype</span><span class="p">,</span> <span class="n">int_kOfKFold</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">),</span> <span class="n">int_nJobs</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="n">int_thread</span><span class="p">))</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Ensemble with co-variate performance (Average of the Pearson and Spearman correlation)&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Training: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_train</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;Testing (&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">k</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;-fold CV): &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">float_score_test</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        
        <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&quot;end analysis at: &quot;</span> <span class="o">+</span> <span class="n">time</span><span class="o">.</span><span class="n">strftime</span><span class="p">(</span><span class="s2">&quot;%Y%m</span><span class="si">%d</span><span class="s2">-%H:%M:%S&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">localtime</span><span class="p">())</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span></div>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Chester (Yu-Chuan Chang)

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>