

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genepi.step3_splitByGene &mdash; GenEpi  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home" alt="Documentation Home"> GenEpi
          

          
          </a>

          
            
            
              <div class="version">
                2.0.10
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../format.html">I/O File Fomats</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../example.html">More Usage Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../workflow.html">How it Works</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API Documentations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">Release History</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">GenEpi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>genepi.step3_splitByGene</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genepi.step3_splitByGene</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Created on Feb 2018</span>

<span class="sd">@author: Chester (Yu-Chuan Chang)</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="c1"># import libraries</span>
<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<span class="c1"># main function</span>
<span class="sd">&quot;&quot;&quot;&quot;&quot;&quot;</span><span class="s2">&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;&quot;</span>
<div class="viewcode-block" id="SplitMegaGene"><a class="viewcode-back" href="../../api.html#genepi.step3_splitByGene.SplitMegaGene">[docs]</a><span class="k">def</span> <span class="nf">SplitMegaGene</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">,</span> <span class="n">int_window</span><span class="p">,</span> <span class="n">int_step</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    In order to extract genetic features for a gene, this function used the start and end positions of each gene from the local UCSC database to split the genetic features. Then, generate the .GEN files for each gene in the folder named snpSubsets.</span>

<span class="sd">    Args:</span>
<span class="sd">        list_snpsOnGene (list): A list contains SNPs on a gene</span>
<span class="sd">        int_window (int): The size of the sliding window</span>
<span class="sd">        int_step (int): The step of the sliding window</span>
<span class="sd">        str_outputFilePath (str): File path of output file</span>
<span class="sd">        str_outputFileName (str): File name of output file</span>

<span class="sd">    Returns:</span>
<span class="sd">        None</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1">### write to gen file if this gene is not mega gene</span>
    <span class="n">int_total_window</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">)</span><span class="o">-</span><span class="n">int_window</span><span class="p">)</span><span class="o">/</span><span class="n">int_step</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">int_total_window</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span> <span class="o">+</span> <span class="s2">&quot;.gen&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_outputFile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_snpsOnGene</span><span class="p">:</span>
                <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>

    <span class="c1">### write gen file of each window on current gene (output file name: geneSymbol_numOfSNPOnGene@windowNum.gen)</span>
    <span class="k">else</span><span class="p">:</span> 
        <span class="k">for</span> <span class="n">idx_w</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">int_total_window</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">idx_w</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">str_outputFileName</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;.gen&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_outputFile</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_snpsOnGene</span><span class="p">[</span><span class="n">int_step</span><span class="o">*</span><span class="n">idx_w</span><span class="p">:</span><span class="n">int_step</span><span class="o">*</span><span class="n">idx_w</span><span class="o">+</span><span class="n">int_window</span><span class="p">]:</span>
                    <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
        
        <span class="c1">### write reminder SNPs to gen file</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">int_total_window</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">str_outputFileName</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;.gen&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_outputFile</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">list_snpsOnGene</span><span class="p">[</span><span class="n">int_step</span><span class="o">*</span><span class="n">int_total_window</span><span class="p">:]:</span>
                <span class="n">file_outputFile</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
    
    <span class="k">return</span></div>

<div class="viewcode-block" id="SplitByGene"><a class="viewcode-back" href="../../api.html#genepi.step3_splitByGene.SplitByGene">[docs]</a><span class="k">def</span> <span class="nf">SplitByGene</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="n">str_inputFileName_UCSCDB</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span> <span class="s2">&quot;/UCSCGenomeDatabase.txt&quot;</span><span class="p">,</span> <span class="n">str_outputFilePath</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>

<span class="sd">    In order to extract genetic features for a gene, this function used the start and end positions of each gene from the local UCSC database to split the genetic features. Then, generate the .GEN files for each gene in the folder named snpSubsets.</span>

<span class="sd">    Args:</span>
<span class="sd">        str_inputFileName_genotype (str): File name of input genotype data</span>
<span class="sd">        str_inputFileName_UCSCDB (str): File name of input genome regions</span>
<span class="sd">        str_outputFilePath (str): File path of output file</span>

<span class="sd">    Returns:</span>
<span class="sd">        - Expected Success Response::</span>

<span class="sd">            &quot;step3: Split by gene. DONE!&quot;</span>
<span class="sd">    </span>
<span class="sd">    Warnings:</span>
<span class="sd">        &quot;Warning of step3: .gen file should be sorted by chromosome and position&quot;</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning of step3: .gen file should be sorted by chromosome and position&quot;</span><span class="p">)</span>

    <span class="n">int_window</span> <span class="o">=</span> <span class="mi">1000</span>
    <span class="n">int_step</span> <span class="o">=</span> <span class="mi">200</span>
    
    <span class="c1">### set default output path</span>
    <span class="k">if</span> <span class="n">str_outputFilePath</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">str_outputFilePath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">),</span> <span class="s2">&quot;snpSubsets&quot;</span><span class="p">)</span>
    <span class="c1">### if output folder doesn&#39;t exist then create it</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">str_outputFilePath</span><span class="p">)</span>
    
    <span class="c1">### load UCSC Genome Database</span>
    <span class="n">list_UCSCGenomeDatabase</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">str_inputFileName_UCSCDB</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_inputFile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_inputFile</span><span class="p">:</span>
            <span class="n">list_UCSCGenomeDatabase</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">))</span>
    <span class="n">np_UCSCGenomeDatabase</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">list_UCSCGenomeDatabase</span><span class="p">)</span>
    
    <span class="c1">### scan all snp</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">str_inputFileName_genotype</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_inputFile</span><span class="p">:</span>
        <span class="n">idx_gene</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">list_snpsOnGene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">file_inputFile</span><span class="p">:</span>
            <span class="c1">### get information of each snp</span>
            <span class="n">list_thisSnp</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
            <span class="n">int_chromosome</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">list_thisSnp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">int_position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">list_thisSnp</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>

            <span class="c1">### current gene is in next chromosome</span>
            <span class="k">if</span> <span class="n">int_chromosome</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                <span class="k">continue</span>
            <span class="c1">### current snp of genotype data is in next chromosome</span>
            <span class="k">elif</span> <span class="n">int_chromosome</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1">#### write gen file of current gene (output file name: geneSymbol_numOfSNPOnGene.gen)</span>
                    <span class="c1">#str_outputFileName = str(np_UCSCGenomeDatabase[idx_gene, 4]) + &quot;_&quot; + str(len(list_snpsOnGene)) + &quot;.gen&quot;</span>
                    <span class="c1">#with open(os.path.join(str_outputFilePath, str_outputFileName), &quot;w&quot;) as file_outputFile:</span>
                    <span class="c1">#    for item in list_snpsOnGene:</span>
                    <span class="c1">#        file_outputFile.writelines(item)</span>
                    <span class="n">str_outputFileName</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">4</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">))</span>
                    <span class="n">SplitMegaGene</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">,</span> <span class="n">int_window</span><span class="p">,</span> <span class="n">int_step</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span><span class="p">)</span>
                <span class="n">list_snpsOnGene</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">while</span> <span class="n">int_chromosome</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                    <span class="c1">### jump to next gene</span>
                    <span class="n">idx_gene</span> <span class="o">=</span> <span class="n">idx_gene</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="c1">### if no next gene then break</span>
                    <span class="k">if</span> <span class="n">idx_gene</span> <span class="o">==</span> <span class="n">np_UCSCGenomeDatabase</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                        <span class="k">break</span>
                    <span class="c1">### current snp on next gene</span>
                    <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">int_position</span> <span class="ow">and</span> <span class="n">int_position</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span> <span class="ow">and</span> <span class="n">int_chromosome</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                        <span class="n">list_snpsOnGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
    
            <span class="c1">### chromosome numbers of current snp and gene are match</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="c1">### current snp on current gene</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">int_position</span> <span class="ow">and</span> <span class="n">int_position</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">2</span><span class="p">]):</span>
                    <span class="n">list_snpsOnGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
                <span class="c1">### snp position exceed this gene</span>
                <span class="k">elif</span> <span class="n">int_position</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">2</span><span class="p">]):</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="c1">#### write gen file of current gene (output file name: geneSymbol_numOfSNPOnGene.gen)</span>
                        <span class="c1">#str_outputFileName = str(np_UCSCGenomeDatabase[idx_gene, 4]) + &quot;_&quot; + str(len(list_snpsOnGene)) + &quot;.gen&quot;</span>
                        <span class="c1">#with open(os.path.join(str_outputFilePath, str_outputFileName), &quot;w&quot;) as file_outputFile:</span>
                        <span class="c1">#    for item in list_snpsOnGene:</span>
                        <span class="c1">#        file_outputFile.writelines(item)</span>
                        <span class="n">str_outputFileName</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">4</span><span class="p">])</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">))</span>
                        <span class="n">SplitMegaGene</span><span class="p">(</span><span class="n">list_snpsOnGene</span><span class="p">,</span> <span class="n">int_window</span><span class="p">,</span> <span class="n">int_step</span><span class="p">,</span> <span class="n">str_outputFilePath</span><span class="p">,</span> <span class="n">str_outputFileName</span><span class="p">)</span>
                    <span class="n">list_snpsOnGene</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">while</span> <span class="n">int_position</span> <span class="o">&gt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span> <span class="ow">and</span> <span class="n">int_chromosome</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                        <span class="c1">### jump to next gene</span>
                        <span class="n">idx_gene</span> <span class="o">=</span> <span class="n">idx_gene</span> <span class="o">+</span> <span class="mi">1</span>
                        <span class="c1">### if no next gene then break</span>
                        <span class="k">if</span> <span class="n">idx_gene</span> <span class="o">==</span> <span class="n">np_UCSCGenomeDatabase</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                            <span class="k">break</span>
                        <span class="c1">### snp on next gene</span>
                        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="n">int_position</span> <span class="ow">and</span> <span class="n">int_position</span> <span class="o">&lt;=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span> <span class="ow">and</span> <span class="n">int_chromosome</span> <span class="o">==</span> <span class="nb">int</span><span class="p">(</span><span class="n">np_UCSCGenomeDatabase</span><span class="p">[</span><span class="n">idx_gene</span><span class="p">,</span> <span class="mi">0</span><span class="p">]):</span>
                            <span class="n">list_snpsOnGene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

            <span class="c1">### if the index of gene out of the boundary of DB then break</span>
            <span class="k">if</span> <span class="n">idx_gene</span> <span class="o">&gt;=</span> <span class="n">np_UCSCGenomeDatabase</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="k">break</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;step3: Split by gene. DONE!&quot;</span><span class="p">)</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Chester (Yu-Chuan Chang)

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>